rs/crypto/src/vetkd/mod.rs

```diff
@@ -0,0 +1,560 @@
+            VetKdKeyShareCreationError::TransientInternalError(error)
+        }
+    }
+}
+
+fn verify_encrypted_key_share_internal<S: CspSigner>(
+    lockable_threshold_sig_data_store: &LockableThresholdSigDataStore,
+    registry: &dyn RegistryClient,
+    csp_signer: &S,
+    key_share: &VetKdEncryptedKeyShare,
+    signer: NodeId,
+    args: &VetKdArgs,
+) -> Result<(), VetKdKeyShareVerificationError> {
+    let registry_version_from_store = lockable_threshold_sig_data_store
+        .read()
+        .transcript_data(&args.ni_dkg_id)
+        .map(|transcript_data| transcript_data.registry_version())
+        .ok_or_else(|| {
+            VetKdKeyShareVerificationError::ThresholdSigDataNotFound(
+                ThresholdSigDataNotFoundError::ThresholdSigDataNotFound {
+                    dkg_id: args.ni_dkg_id.clone(),
+                },
+            )
+        })?;
+
+    let signature = BasicSigOf::new(BasicSig(key_share.node_signature.clone()));
+    BasicSigVerifierInternal::verify_basic_sig(
+        csp_signer,
+        registry,
+        &signature,
+        &key_share.encrypted_key_share,
+        signer,
+        registry_version_from_store,
+    )
+    .map_err(VetKdKeyShareVerificationError::VerificationError)
+}
+
+fn combine_encrypted_key_shares_internal<C: ThresholdSignatureCspClient>(
+    lockable_threshold_sig_data_store: &LockableThresholdSigDataStore,
+    threshold_sig_csp_client: &C,
+    logger: &ReplicaLogger,
+    shares: &BTreeMap<NodeId, VetKdEncryptedKeyShare>,
+    args: &VetKdArgs,
+) -> Result<VetKdEncryptedKey, VetKdKeyShareCombinationError> {
+    let transcript_data_from_store = lockable_threshold_sig_data_store
+        .read()
+        .transcript_data(&args.ni_dkg_id)
+        .cloned()
+        .ok_or_else(|| {
+            VetKdKeyShareCombinationError::ThresholdSigDataNotFound(
+                ThresholdSigDataNotFoundError::ThresholdSigDataNotFound {
+                    dkg_id: args.ni_dkg_id.clone(),
+                },
+            )
+        })?;
+    let pub_coeffs_from_store = match transcript_data_from_store.public_coefficients() {
+        PublicCoefficients::Bls12_381(pub_coeffs) => &pub_coeffs.coefficients,
+    };
+    let reconstruction_threshold = pub_coeffs_from_store.len();
+    let master_public_key = {
+        let public_key_bytes = pub_coeffs_from_store
+            .iter()
+            .copied()
+            .next()
+            .ok_or_else(|| {
+                VetKdKeyShareCombinationError::InternalError(format!(
+                    "failed to determine master public key: public coefficients for NI-DKG ID {} are empty",
+                    &args.ni_dkg_id
+                ))
+            })?;
+        ic_crypto_internal_bls12_381_vetkd::G2Affine::deserialize(&public_key_bytes).map_err(
+            |_: PairingInvalidPoint| VetKdKeyShareCombinationError::InvalidArgumentMasterPublicKey,
+        )?
+    };
```

Comment:
This same sequence occurs twice (here and `verify_encrypted_key_internal`) can we extract it to a single helper `fn` to reduce the duplication?